### PR TITLE
Ghost command now releases artifact ghost role

### PR DESCRIPTION
**Pull request by osjarw**
## About the PR
If someone picks the sentient artifact ghost role and starts to regret it, they can now `/ghost` and let someone else take the role.

## Why / Balance
Before this PR, the sentient artifact ghost role couldn't be released 

## Technical details
`CanReturn` is set to true because the artifact doesn't have a mobstate, so `IsCharacterDeadPhysically` returns true. Because of this, `/ghost` doesn't release the role.
~I added an if clause that sets `CanReturn` to false when the artifact/player/mind has the `ArtifactComponent`, which implies that this is a sentient artifact ghost role.~
Added a event listener on `GhostAttemptHandleEvent` which calls `OnGhostAttempt` with parameters which don't allow the player to return to the body, which frees up the ghost role.

## Media

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

Before:
The role can't be selected after the artifact has ghosted

https://github.com/space-wizards/space-station-14/assets/62134478/2ea6b33d-1dac-4b52-b29e-c6ce1cc5b1b2

After:
The role can now be selected after the artifact has ghosted

https://github.com/space-wizards/space-station-14/assets/62134478/1d67ad6a-d529-42a9-b6d1-ca4bc6fc17ae

## Breaking changes

**Changelog**

:cl:
- fix: Ghost command now frees up sentient artifact ghost role

### CODE REVIEW

File: Content.Server/GameTicking/GameTicker.GamePreset.cs [hunk 1]
```diff
@@ -274,6 +275,10 @@ public bool OnGhostAttempt(EntityUid mindId, bool canReturnGlobal, bool viaComma
                 }
             }
 
+            // Free the sentient artifact ghost role when ghosting
+            if(canReturnGlobal && TryComp(playerEntity, out ArtifactComponent? artifact))
+                canReturn = false;
+
```
Comment (metalgearsloth) [inline Content.Server/GameTicking/GameTicker.GamePreset.cs:281]: Artifacts shouldn't be hardcoded into ghost code as it just becomes a big mess for forks, probably put in a bool on the component instead or use an event.

**Comment by AJCM-git** [comment]
@osjarw you coming back to this?

**Comment by osjarw** [comment]
I can take a look at this again later today.

**Comment by osjarw** [comment]
Working on this currently, managed to hack together a new way to solve this issue, implemented with events. 
Currently I just need to figure out how to solve the merge conflict properly and just push it.

**Comment by Mirino97** [comment]
Reopened by contributor's request

**Comment by osjarw** [comment]
Hopefully this is an acceptable way of fixing this.. 
I guess that I could also just go and figure out the "proper" way of doing this ghost stuff without just re-calling OnGhostAttempt. 
Anyways also tested this code, it seems to be working and it didn't seem to break anything, like ghosting while in crit or when playing as a mouse. 

Also sorry for taking this long in submitting this code, been a bit busier than expected these past few weeks.

**Comment by osjarw** [comment]
It looks like that test is failing on the code I submitted, I guess I'll try to figure out how to make it run properly, without failing that test...

**Comment by osjarw** [comment]
While this PR is in a "mergeable" state, I think it should actually handle the ghost attempt. 
Currently it just calls `OnGhostAttempt` again, and now that I've taken a closer look at it, it isn't really something that should be done. The logging happens twice, and there's a tiny bit of recursion, which isn't really ideal. 

I guess I could just yank out the important bits from `OnGhostAttempt`, but that would cause an annoying amount of code duplication, and git blame wouldn't look as beautiful as it does now..

Actually I guess I'll just go with the recursion route, I'll just set `viaCommand` to false to stop the logging from happening twice, and actually do some null-checks instead of just assuming that the ones that are already done apply to my code.

**Comment by metalgearsloth** [comment]
It's still failing tests.

**Comment by osjarw** [comment]
Finally got around to implementing what I wrote in my last comment..

The `DeleteAllThenGhost` test shouldn't be failing anymore, used to fail because one of the nullables was null and I assumed it to already be checked... 
TIL that TryComp can be called with a nullable, in which case it does the handling, so switched to using that. 

Also the admin logs aren't duplicated like I thought in my last comment, apparently I fixed that already.

**Comment by osjarw** [comment]
This should close #28081

**Comment by AJCM-git** [comment]
Isnt this already handled by the `GhostRoleComponent.ReregisterOnGhost` field? prob should just fix that if its somehow broken

**Review by AJCM-git — CHANGES_REQUESTED** [review]
What i said before, forgot to leave it as a review

**Pull request closed without merging**